### ISSUE TITLE
iq response tracker doesn't clean up

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2009]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)| @ralphm reported|
|-|-|
|Trac ID|trac#2009|
|Type|defect|
|Created|2006-08-21 15:39:10Z|

Two issues:

* On callback or errback, the deferred is not removed from the iqDeferreds dictionary.
* On stream disconnect, the deferreds can never be fired anymore.

The first issue is simply because of a bad implementation.
The second issue can be avoided by errbacking all deferreds with a `twisted.internet.error.ConnectionLost` failure.

<details><summary>Searchable metadata</summary>

```
trac-id__2009 2009
type__defect defect
reporter__ralphm ralphm
priority__highest highest
milestone__Words_0_5 Words-0.5
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1156174750000000 1156174750000000
changetime__1156193121000000 1156193121000000
version__None None
owner__ 
cc__exarkun cc__ralphm
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_1"></a>|@ralphm removed owner|
|-|-|

Implemented in [branches/jabber-iqtracker-cleanup-2009.](branches/jabber-iqtracker-cleanup-2009.) Please review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to @ralphm|
|-|-|

This branch generally looks good.  Good tests, nice docstrings.  One minor aesthetic point - having a leading single space in docstrings is kind of ugly.  Instead of (for example):

```
""" Handle iq response by firing associated deferred. """
```

and

```
""" Make sure deferreds do not linger on after disconnect.

...
"""
```

How about

```
"""
Handle iq response by firing associated deferred.
"""
```

and

```
"""
Make sure deferreds do not linger on after disconnect.

...
"""
```

Aside from that, the body of ```disconnected``` looks a little dangerous.  Is it possible for the errbacks on any of the ```iqDeferreds``` to run any code which might try to change ```xs.iqDeferreds```?  If so, this will blow up part way through the loop and prevent all the Deferreds from being errback'd.  If there's any chance of this occurring (and ```xs.iqDeferreds``` being reset to an empty dictionary seems to indicate that there is), I'd recommend rewriting the loop like this:

```
    iqDeferreds = xs.iqDeferreds
    xs.iqDeferreds = {}
    for d in iqDeferreds.itervalues():
        d.errback(ConnectionLost())
```

If you make this change, it would be nice to have a test for this behavior as well.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_3"></a>|@ralphm set owner to @exarkun|
|-|-|

I added a test to guard against the blow up and rewritten the loop as you suggested. Also changed the comments. Old habits die hard. Please review again.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @ralphm|
|-|-|

Looks good!  Merge away.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_5"></a>|@ralphm set status to `closed`|
|-|-|

(In [17910]) Make iq tracker clean up after disconnect or firing the associated deferred.

Author: ralphm Reviewer: exarkun Fixes [#2009](https://github.com/twisted/twisted/issues/2009)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
